rfcs/0168-Trigger-Tests-Based-on-PR-Comments.md

```diff
@@ -0,0 +1,31 @@
+# RFC 168 - Allow comments to trigger tasks for non-collaborators
+* Comments: [#168](https://github.com/taskcluster/taskcluster-rfcs/pull/168)
+* Proposed by: @bhearsum
+
+# Summary
+
+Allow collaborators to a Github repository to request that Tasks be run on Pull Requests that ran no tasks due to a restrictive `pullRequests` policy.
+
+## Motivation
+
+Increasingly, Taskcluster is being used for projects hosted on Github. In many of these projects, Tasks that are run during Pull Requests rely on secrets or other sensitive information. In these cases we usually use a `pullRequests` policy of `collaborators` to ensure that untrusted people cannot gain access to such things. Because of this, it means we have no reasonable way of verifying Pull Requests submitted by non-collaborators. Workarounds have been found (eg: a separate Github Actions flow for such PRs) but this is both burdensome and typically does not provide robust enough testing.
+
+# Details
+
+`.taskcluster.yml` will be modified to support a new `allowApprovals` policy, which will support `collaborators` as a value. When set, collaborators to the repository may add a comment containing the string "taskcluster go" to a Pull Request that hasn't run tests because of its `pullRequests` policy. When this is done, Taskcluster will run any Tasks that it would've ran when the Pull Request was opened, had the author been authorized to start Tasks.
```

Comment:
I'm not 100% sure about the string `taskcluster go`; we may want to look at bors and other comment-based triggers to see what syntax is common. But other than that, 👍

Reply:
I probably should've said that this was mostly a placeholder string. I have no strong feelings on what the actual string is.

Reply:
Could this string be passed to the JSON-e and allowed to do whatever it would like there?

So:

```
/taskcluster something here
```

results in `.taskcluster.yml` being rendered with a different `tasks_for` and the value "something here" passed in the JSON-e context.  Then it's up to `.taskcluster.yml` to decide what to do with it.  In simple cases it could just ignore the value and run the same stuff it would on a PR.  But in more complex cases (such as that for @marco-c) it could do more interesting things.

Maybe I'm over-generalizing.  It seems like there's some middle-ground that would be flexible enough to handle more than just "do what you would have done if the submitter was a collaborator" without introducing a lot of additional complexity of implementation or difficulty of usage.

Reply:
Based on my read of the tc-gh code, that seems pretty reasonable doable (AFAICT we would just populate the message sent to the pulse queue with the same thing as `pull_request`, except a different `tasks_for`). And if I understand correctly, this should be a simple change in `.taskcluster.yml` (check for `github-pull-request` and `github-some-magic-string` in places where we're currently checking just `github-pull-request`)

Reply:
I was thinking [issue_comment](https://docs.github.com/en/developers/webhooks-and-events/webhook-events-and-payloads#issue_comment) could be another event that .taskcluster.yml could handle, but I guess that would be harder to implement and maybe too complex to use.

Reply:
I may be misunderstanding what you're saying -- but that's what this proposal is already saying. More concretely, to make sure we're on the same page, I think we'd want a new block around https://github.com/taskcluster/taskcluster/blob/11aca1d967c86b1a4b565e6ba1fbf8bf86166f2c/services/github/src/api.js#L240 for `issue_comment`, that populates data as described in my previous comment.

Reply:
What I was thinking about is: .taskcluster.yml rendered with the `issue_comment` event, and so tasks to run defined directly via the JSON-e rendering through if conditions.
From your proposal I (wrongly) thought you wanted to run, on `issue_comment` events, the tasks that would have run on the PR for the `github-pull-request` event, and not let the projects' .taskcluster.yml directly handle the `issue_comment` event.

Reply:
Oh, I see. Right, the proposal was written before the idea of sending a different `tasks_for` to .taskcluster.yml was thrown out. I think what Dustin is proposing is similar to what you're thinking (although a bit more rigid, because you wouldn't get the full `issue_comment` event, just part of the actual comment as part of the `tasks_for` string.

Reply:
You've characterized my suggestion correctly.  My rigidity was just trying to limit the generality of the idea.  Providing the full event for every comment is certainly possible, but implementing logic like "does this comment have the text `taskcluster go` in it" or `does this comment have a command of the form `/run-suite <suiteName>`" in JSON-e is rather difficult!  The wrinkle is, the security model will depend on whether the person making the comment is a collaborator or not, which is definitely not something that can be calculated in JSON-e.

I think it's OK for this proposal to be a little more ad-hoc and limited in its scope.  Bonus points if that doesn't preclude a later, more general solution.

Reply:
I agree that it's probably best not to focus on a fully generalized solution here. However, @marco-c, what information do you need to address your specific use case? It could be that we could add the specific things needed for the known cases now (if they're relatively simple/easy), and evolve later - as @djmitche suggests.

Reply:
The more rigid solution proposed by Dustin actually works for my purposes.

Reply:
Great. I'm going to update this to reflect that - it seems like the right balance for now, based on the known use cases.

Reply:
>  The wrinkle is, the security model will depend on whether the person making the comment is a collaborator or not, which is definitely not something that can be calculated in JSON-e.

@djmitche - I was assuming that Taskcluster-Github would do this check (as it does for the `pullRequests` policy), but this comment seems to hint otherwise. Can you clarify?

Reply:
I also want to clarify whether or not you think the comment should go in `tasks_for`, or a separate variable. Eg: if I commented with `/taskcluster go` should that result in `tasks_for` being set to `github-comment-go` or `github-comment`, and a new variable (let's say `event.comment`) having `go` in it?

Reply:
> @djmitche - I was assuming that Taskcluster-Github would do this check (as it does for the `pullRequests` policy), but this comment seems to hint otherwise. Can you clarify?

Taskcluster-github will have to do make this check, yes.  I think the choices are either to do that check and not render anything in the case of a comment by someone not matching the policy; or to do the check and then provide the result in the JSON-e context (`{"poster-is-allowed": true}`?).  I'd be OK with either of those options -- or something else!  I have a mild preference for the first option just because it matches the approach for PRs.

> I also want to clarify whether or not you think the comment should go in `tasks_for`, or a separate variable. Eg: if I commented with `/taskcluster go` should that result in `tasks_for` being set to `github-comment-go` or `github-comment`, and a new variable (let's say `event.comment`) having `go` in it?

String manipulation in JSON-e is tricky, so I think a separate variable would be best.

Reply:
I'm fully on board with your preferences here -- it seems the most clean and straightforward. Thanks for the clarifications!

Reply:
I just pushed a change that I _think_ encapsulates all of this.